app/controllers/cloud_volume_controller.rb

```diff
@@ -620,6 +683,23 @@ def delete_cloud_volumes(volumes)
                  volumes.length) % {:number => volumes.length})
   end
 
+  def safe_delete_cloud_volumes(volumes)
```

Comment:
I'm pretty sure this should be implemented in the API repo if it's not yet.

Reply:
@skateman, the safe_delete is implemented similar to the regular delete. safe_delete_cloud_volumes() is a modified clone of delete_cloud_volumes() which already exist inside the cloud_volume_controller. can you please clarify what need to be in the API repo and what in the controller ?

Reply:
@llivne AFAIK there's a way to directly call API endpoints from the toolbars, I need to do a research arond it, but if it doesn't exist, then this might be the right moment to implement it. We are trying to (slowly) convert this repo into a single page js-only API-driven application and by introducing new controller actions we are taking steps backwards. I will do some research about this and get back to you, in general the solution should be simpler than the current implementation for sure.

Would be the non-safe delete used on the cloud volume types you are introducing? With other words: will be both delete action exposed to a single cloud volume? I'm asking because we could just override the regular `delete` method in your provider model and let the OO inheritance decide which one to use depending on what provider are you deleting. This way you could apply the operation to multiple cloud volumes coming from different providers at the same time and we would not even care which one gets called, because the model inheritance just sorts it out by its own magic. I guess this is how our providers supposed to work, abstract away all the provider-specific stuff and expose common methods, cc @agrare. Of course this is only doable if you don't want the "unsafe" delete on your entities, only the safe.

The [`delete`](https://github.com/ManageIQ/manageiq-api/blob/master/config/api.yml#L751-L752) operation is already implemented on the cloud volume API endpoint, so if we go with the above solution, you don't need to do anything in the API...

Reply:
So I found one, [here](https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/helpers/application_helper/toolbar/ansible_repositories_center.rb#L50-L66) you have a nice example of calling an API endpoint from the toolbar to delete an entity and then even redirect to a page or refresh the current one to take effect.

Reply:
@skateman, using API seems much simpler, I will learn on it and see how to address this issue. we're going to split PR-7282 to several small PRs (hopefully this week)  so the safe-delete will get its own PR and it will be simpler to see the changes in there. I do think that if regular delete is implemented with controller action, the safe delete should be implemented in the same manner, and both should be updated to API-driven application together. I will gladly do both modifications but since we are on a tight schedule is it ok to do it in two steps ? (first introducing safe delete via controller action, and next changing both safe delete and regualar delete to work via API) 

As for your other question, we intend to keep both options open for the new provider (regular delete and safe delete).

Reply:
Okay, so if both methods are necessary, it's a little bit more complex. The thing I'd like to ask here is to move this method out to the API (I can help you with that or even implement it for you in a few minutes). Then you need to change ~3 lines in the toolbar button's definition and this method can be deleted. This is one way to make your huge PR less complex and it's definitely easy to review and get merged a single method PR in the API.

Reply:
@skateman if you can help with the API it will be great ( I'm not familiar with the API repo and I guess it will take me much longer then you to implement it). I can handle the changes and PR in the API repo if you simply redirect me to where to add the code, or for a relevant example. I looked on the changes needed in the toolbar and like you said, once we have the API it quiet simple.

Reply:
@llivne I'll send you an API PR probably tomorrow, today is full already :confused:

Reply:
great, thanks!

Reply:
Here you go: https://github.com/ManageIQ/manageiq-api/pull/895

You only need to implement the mentioned method in your model and the API endpoint would be:
```
POST /api/cloud_volumes/XY/safe_delete
```

Reply:
awesome!

Reply:
@skateman, I worked on moving the safe_delete to work via API but I have some concerns about it. All of the  cloud volumes action buttons (e.g. create, snapshot, attach, backup ....) are done with the controller, and I'm not sure it wise to distinguish only the safe_delete to work with the API. I think all the actions should work with the same mechanism. Also, action performed via the controller has an additional validation step to check that it allowed, and this is something that need to be implemented if we use API. I really think that this is something that should be done in two steps, first adding the safe_delete just like all other buttons, and then changing all of the cloud volume buttons to work with API (I understand from your comments that this is the preferred way by miq).

it may be easier to talk about this subject in a separate PR (I hoped to open it by now, but we had some problems with the split, hopefully all be done soon).

liran

Reply:
@llivne, if we copy-paste bad code patterns, we will never achieve our goal of converting our frontend into an API-driven single page React application. The toolbars have been made declarative exactly to allow this kind of gradual conversion of controller actions to API endpoints. The RBAC validation is also implemented in the API and any additional failed checks should be able to raise errors that the frontend is capable of rendering.

The right solution preferred by miq would be to first convert the controller actions to API endpoints and then implement your new controller action in the API. However, as you said you're on a tight schedule, it's okay to have your action in the API and move the rest of them in a followup PR.